common/matrix/csr_kernels.hpp.inc

```diff
@@ -0,0 +1,793 @@
+    const IndexType ind_end = end - wsize;
+    const auto tile_block =
+        group::tiled_partition<wsize>(group::this_thread_block());
+    for (; ind < ind_end; ind += wsize) {
+        process_window<false>(tile_block, num_rows, data_size, ind, &row,
+                              &row_end, &nrow, &nrow_end, &temp_val, val,
+                              col_idxs, row_ptrs, b, b_stride, c, c_stride,
+                              column_id, scale);
+    }
+    process_window<true>(tile_block, num_rows, data_size, ind, &row, &row_end,
+                         &nrow, &nrow_end, &temp_val, val, col_idxs, row_ptrs,
+                         b, b_stride, c, c_stride, column_id, scale);
+    warp_atomic_add(tile_block, true, &temp_val, row, c, c_stride, column_id,
+                    scale);
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(spmv_block_size) void abstract_spmv(
+    const IndexType nwarps, const IndexType num_rows,
+    const ValueType *__restrict__ val, const IndexType *__restrict__ col_idxs,
+    const IndexType *__restrict__ row_ptrs, const IndexType *__restrict__ srow,
+    const ValueType *__restrict__ b, const size_type b_stride,
+    ValueType *__restrict__ c, const size_type c_stride)
+{
+    spmv_kernel(nwarps, num_rows, val, col_idxs, row_ptrs, srow, b, b_stride, c,
+                c_stride, [](const ValueType &x) { return x; });
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(spmv_block_size) void abstract_spmv(
+    const IndexType nwarps, const IndexType num_rows,
+    const ValueType *__restrict__ alpha, const ValueType *__restrict__ val,
+    const IndexType *__restrict__ col_idxs,
+    const IndexType *__restrict__ row_ptrs, const IndexType *__restrict__ srow,
+    const ValueType *__restrict__ b, const size_type b_stride,
+    ValueType *__restrict__ c, const size_type c_stride)
+{
+    ValueType scale_factor = alpha[0];
+    spmv_kernel(nwarps, num_rows, val, col_idxs, row_ptrs, srow, b, b_stride, c,
+                c_stride, [&scale_factor](const ValueType &x) {
+                    return scale_factor * x;
+                });
+}
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size) void set_zero(
+    const size_type nnz, ValueType *__restrict__ val)
+{
+    const auto ind =
+        static_cast<size_type>(blockDim.x) * blockIdx.x + threadIdx.x;
+    if (ind < nnz) {
+        val[ind] = zero<ValueType>();
+    }
+}
+
+
+template <typename IndexType>
+__forceinline__ __device__ void merge_path_search(
+    const IndexType diagonal, const IndexType a_len, const IndexType b_len,
+    const IndexType *__restrict__ a, const IndexType offset_b,
+    IndexType *__restrict__ x, IndexType *__restrict__ y)
+{
+    auto x_min = max(diagonal - b_len, zero<IndexType>());
+    auto x_max = min(diagonal, a_len);
+    while (x_min < x_max) {
+        auto pivot = (x_min + x_max) >> 1;
```

Comment:
I just noticed this bug: The calculation of `x_min + x_max` can overflow, and since our `IndexType` is always(?) signed, the resulting midpoint is negative, as the shift operator keeps the sign of its operand. [Source](https://en.cppreference.com/w/cpp/language/operator_arithmetic#Bitwise_shift_operators)
More details: [This blog post](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html) explains it nicely.

```suggestion
        auto pivot = x_min + (x_max - x_min) / 2;
```